api/reporting/schemas/reporting-codegen.rb

```diff
@@ -136,68 +261,190 @@ def to_swagger_property(column)
+    + COLUMNS.map do |field|
+      "+ \"  #{adjust_col(field[:name])}"
     end \
     .join(",\\n\"\n") \
     + "\\n\"\n" \
-    + "+ \"FROM Db#{to_camel_case(table_name, true)} #{table_alias}\")\n" \
-    + "  List<#{projection_name(table_name)}> getReporting#{to_camel_case(table_name, true)}s();"
+    + "+ \"FROM #{TABLE_INFO[:entity_class]} #{TABLE_INFO[:sql_alias]}\")\n" \
+    + "  List<#{TABLE_INFO[:projection_interface]}> getReporting#{to_camel_case(TABLE_INFO[:name], true)}s();"
+end
+
+sql = to_query
+
+write_output(OUTPUTS[:projection_query], sql, 'Projection Query')
+
+# Unit Test Constants
+#
+BASE_TIMESTAMP = Time.new(2015, 5, 5).freeze
+TIMESTAMP_DELTA_SECONDS = 24 * 60 * 60 # .freeze # seconds in day
+
+# N.B. some Short fields are only valid up to the number of associated enum values - 1. Fixing these
+# up by hand for now.
+def to_constant_declaration(column, index)
+  value = case column[:java_type]
+          when 'String'
+            "\"foo_#{index}\""
+          when 'Integer'
+            "%d" % [index]
+          when 'Long'
+            "%dL" % [index]
+          when 'Double'
+            "%f" % [index + 0.5]
+          when 'Boolean'
+            index.even? # just flip it every time
+          when 'Timestamp'
+            # add a day times the index to base timestamp
+            timestamp = BASE_TIMESTAMP + TIMESTAMP_DELTA_SECONDS * index
+            "Timestamp.from(Instant.parse(\"#{timestamp.strftime("%Y-%m-%dT00:00:00.00Z")}\"))"
+          else
+            if column[:is_enum]
+              column[:default_enum]
+            else
+              index.to_s
+            end
+          end
+  "public static final #{column[:java_type]} #{column[:java_constant_name]} = #{value};"
+end
+
+constants = COLUMNS.enum_for(:each_with_index) \
+  .map { |col, index| to_constant_declaration(col, index) } \
+  .join("\n")
+
+write_output(OUTPUTS[:unit_test_constants], constants, 'Unit Test Constants')
+
+### Mock Instantiation
+# Mock the projection interface for testing with mock services exposing them
+mocks = COLUMNS.map { |col|
+  "doReturn(#{col[:java_constant_name]}).when(#{TABLE_INFO[:mock]}).#{col[:getter]}();"
+}
+
+lines = []
+lines << "final #{TABLE_INFO[:projection_interface]} #{TABLE_INFO[:mock]} = mock(#{TABLE_INFO[:projection_interface]}.class);"
+lines << mocks
+lines.flatten!
+
+write_output(OUTPUTS[:unit_test_mocks], lines.join("\n"), 'Unit Test Mocks')
+
+### Assertions
+dto_assertions = COLUMNS.map{ |col|
+  getter_call = "#{to_camel_case(table_name, false)}.#{col[:getter]}()"
+  expected = col[:java_constant_name]
+  if col[:java_type].eql?('Timestamp')
+    "    assertTimeApprox(#{getter_call}, #{expected});"
+  else
+    "    assertThat(#{getter_call}).isEqualTo(#{expected});"
+  end
+
+}.join("\n")
+
+write_output(OUTPUTS[:dto_assertions], dto_assertions, 'Unit Test DTO Assertions')
+
+### Parameter Column Enum
+def object_value_function(col)
```

Comment:
There are two different ultimate output types depending on the uploading mode. For streaming (fast but possibly allows duplicates), we have provide a `RowToInsert` with a map from column name to all non-null values passed as Objects. There are some surprises here, such as a special way you have to format the timestamps.

The QueryParameterValue structure is more familiar, as we use it for select queries from BQ. It's also difficult to work with and error prone, so I've beefed up the `QueryParamterValues` utility class and added testing to it. As written, we do a static import, which makes sense when calling the same method more than 10 times in quick succcession.